r/src/array.cpp

```diff
@@ -373,6 +373,75 @@ std::shared_ptr<arrow::Array> Int64Array(SEXP x) {
   return std::make_shared<typename TypeTraits<Int64Type>::ArrayType>(data);
 }
 
+inline int difftime_unit_multiplier(SEXP x) {
+  std::string unit(CHAR(STRING_ELT(Rf_getAttrib(x, symbols::units), 0)));
+  if (unit == "secs") {
+    return 1;
+  } else if (unit == "mins") {
+    return 60;
+  } else if (unit == "hours") {
+    return 3600;
+  } else if (unit == "days") {
+    return 86400;
+  } else if (unit == "weeks") {
+    return 604800;
+  }
+  Rcpp::stop("unknown difftime unit");
+  return 0;
+}
+
+std::shared_ptr<arrow::Array> Time32Array_From_difftime(SEXP x) {
+  // number of seconds as a double
+  auto p_vec_start = REAL(x);
+  auto n = Rf_xlength(x);
+  int64_t null_count = 0;
+
+  int multiplier = difftime_unit_multiplier(x);
+  std::vector<std::shared_ptr<Buffer>> buffers{nullptr, nullptr};
+
+  R_ERROR_NOT_OK(AllocateBuffer(n * sizeof(int32_t), &buffers[1]));
+  auto p_values = reinterpret_cast<int32_t*>(buffers[1]->mutable_data());
+
+  R_xlen_t i = 0;
+  auto p_vec = p_vec_start;
+  for (; i < n; i++, ++p_vec, ++p_values) {
+    if (NumericVector::is_na(*p_vec)) {
+      break;
+    }
+    *p_values = static_cast<int32_t>(*p_vec * multiplier);
+  }
+
+  if (i < n) {
```

Comment:
I have seen this pattern in some other conversion functions. Seems like you could introduce some lambdas / functors and have a reusable template for this part

Reply:
Yes, there is some repetition. I'll write something more generic, but I can do that in a follow up.